### PR TITLE
fix: FPLA-2726 Fix Judicial messages issues

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2726

### Change description ###
Prepopulate sender and recipient email addresses.
Add validation to check if sender and recipient email addresses are the same in the reply journey.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java [hunk 1]
```diff
@@ -81,6 +83,21 @@
...
 
+    public List<String> validateJudgeReplyMessage(CaseData caseData) {
+        List<String> errors = new ArrayList<>();
+
+        MessageJudgeEventData messageJudgeEventData = caseData.getMessageJudgeEventData();
+        JudicialMessage judicialMessageReply = messageJudgeEventData.getJudicialMessageReply();
+
+        if (judicialMessageReply != null && YES.getValue().equals(judicialMessageReply.getIsReplying())
+            && isNotEmpty(judicialMessageReply.getReplyFrom())
+            && judicialMessageReply.getReplyFrom().equals(judicialMessageReply.getReplyTo())) {
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java:94]: minor, I wonder if we could break this up into smaller more readable private methods e.g

`if (isReplyingToMessge(judicialMessageReply) && hasMatchingReplyEmaiAddress(judicialMessageReply))`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java [hunk 2]
```diff
@@ -205,14 +224,16 @@ public String getNextHearingLabel(CaseData caseData) {
...
                     JudicialMessage updatedMessage = judicialMessage.toBuilder()
                         .updatedTime(time.now())
-                        .sender(sender) // Get the email of the current user
-                        .recipient(judicialMessage.getSender()) // Get the sender of the previous message
+                        .sender(sender)
+                        .recipient(judicialMessageReply.getReplyTo())
                         .messageHistory(buildMessageHistory(judicialMessageReply, judicialMessage, sender))
                         .latestMessage(judicialMessageReply.getLatestMessage())
+                        .replyFrom(null)
+                        .replyTo(null)
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java:236]: Can these be removed?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeServiceTest.java [hunk 3]
```diff
@@ -222,35 +235,49 @@ void shouldInitialiseJudicialMessagesWithOpenStatusWhenC2DocumentsDoNotExist() {
...
+            "judicialMessageMetaData", JudicialMessageMetaData.builder()
+                .sender(COURT_EMAIL)
+                .recipient(EMPTY).build()
         );
 
         assertThat(data).isEqualTo(expectedData);
     }
 
     @Test
     void shouldReturnEmptyMapWhenC2DocumentsDoNotExist() {
```
Comment (Jamesnelson117) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeServiceTest.java:248]: Test name is no longer returning empty map so it's worth updating test name.

File: ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json [hunk 4]
```diff
@@ -1,4 +1,13 @@
 [
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "JudicialMessageMetaData",
+    "CaseEventID": "messageJudgeOrLegalAdviser",
+    "CaseFieldID": "judicialMessageMetaData",
+    "ListElementCode": "sender",
```
[the remark this answers is not in the document]
Comment (aliveni) [reply ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json:7]: `requestedBy` is now used for `subject`. see [here](https://github.com/hmcts/fpl-ccd-configuration/blob/4d9057d30b066f8a5bb07b77f8fbb912f1e4d9fb/service/src/main/java/uk/gov/hmcts/reform/fpl/model/judicialmessage/JudicialMessageMetaData.java#L16)

File: ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json [hunk 5]
```diff
@@ -1,4 +1,13 @@
...
+    "FieldDisplayOrder": 1,
```
[the remark this answers is not in the document]
Comment (aliveni) [reply ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json:8]: updated

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/MessageJudgeController.java [hunk 6]
```diff
@@ -77,6 +77,17 @@ public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody Callback
         return respond(caseDetailsMap);
     }
 
+    @PostMapping("/validate/mid-event")
+    public AboutToStartOrSubmitCallbackResponse validateReviewDecision(@RequestBody CallbackRequest callbackRequest) {
```
[the remark this answers is not in the document]
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/MessageJudgeController.java]: good spot :) renamed now

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java (hunk 2 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java]: yeah, removed

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeServiceTest.java (hunk 3 above)
Comment (aliveni) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeServiceTest.java:248]: renamed the test

File: ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json:7]: In that case let's add a comment via the `Comment` property.

[the remark this answers is not in the document]
Comment (aliveni) [reply ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json:7]: ok, I'll add a comment to say it is `From email address`. is that ok?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java [hunk 7]
```diff
@@ -81,6 +83,19 @@
         return data;
     }
 
+    public List<String> validateJudgeReplyMessage(CaseData caseData) {
+        List<String> errors = new ArrayList<>();
+
+        MessageJudgeEventData messageJudgeEventData = caseData.getMessageJudgeEventData();
+        JudicialMessage judicialMessageReply = messageJudgeEventData.getJudicialMessageReply();
+
+        if (isReplyingToMessage(judicialMessageReply) && hasMatchingReplyEmaiAddress(judicialMessageReply)) {
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java:92]: much nicer :)

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeServiceTest.java (hunk 3 above)
Comment (Jamesnelson117) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeServiceTest.java:248]: Are you sure you did? I do not see it.

File: ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (aliveni) [reply ccd-definition/CaseEventToComplexTypes/messageJudge/judicialMessageMetaData.json:7]: As discussed, added comment on `requestedBy` field

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeServiceTest.java (hunk 3 above)
Comment (aliveni) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeServiceTest.java]: sorry, it was missing earlier. it's updated now

File: ccd-definition/ComplexTypes/CareSupervision/JudicialMessage/JudicialMessageMetaData.json [hunk 8]
```diff
@@ -1,10 +1,18 @@
 [
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "JudicialMessageMetaData",
+    "ListElementCode": "sender",
+    "FieldType": "Email",
+    "ElementLabel": "From email address",
```
[the remark this answers is not in the document]
Comment (aliveni) [reply ccd-definition/ComplexTypes/CareSupervision/JudicialMessage/JudicialMessageMetaData.json]: I checked with Collette. She suggested to update the `From` and `To` email address fields to `Sender's email address` and `Recipient’s email address` fields on both new message and reply journeys

**Review by Jamesnelson117 — APPROVED** [review]
LGTM

**Pull request closed without merging**